csharp: Updated `Nucleotide count` notes (#1118)
exercism/website-copy: tracks/csharp/exercises/nucleotide-count/mentoring.md

### Common Suggestions

> As exercise text doesn't explicitly mention that student should optimize for performance, don't discard solutions just on this ground. Rather use it as a chance to start discussion about readability/performance trade-offs and when to prefer which.

Reviewer: ```suggestion
As the README doesn't explicitly mention that student should optimize for performance, don't discard solutions just on this ground. Rather, use it as a chance to start a discussion about readability/performance trade-offs and when to prefer which.
```

> As the README doesn't explicitly mention that student should optimize for performance, don't discard solutions just on this ground. Rather, use it as a chance to start a discussion about readability/performance trade-offs and when to prefer which.

Reviewer: ```suggestion
As the README doesn't explicitly mention that a student should optimize for performance, don't discard solutions just on this ground. Rather, use it as a chance to start a discussion about readability/performance trade-offs and when to prefer which.
```

> * If you look for perfomance oriented solution iterating over the input string more than once wastes cycles, memory, etc. Often it isn't obvious that this is happening because methods like `ToList`, `GroupBy` or `Count` will iterate a list without the student having to write a loop.

Reviewer: ```suggestion
* If you look for a performance-oriented solution, iterating over the input string more than once wastes cycles, memory, etc. Often, it isn't obvious that this is happening because methods like `ToList`, `GroupBy` and `Count` will iterate over all the string's characters without the student having to write a loop.
```

>   * Hence, using `IEnumerable.Any()` or `IEnumerable.Where(...)` to implement the exception is wrong. It should be thrown within the single iteration.

Reviewer: > It should be thrown within the single iteration.

Well, it is only "wrong" when the student wants to favor performance. Personally, I'm not sure this comment is that useful. @mikedamay what do you think?

Reviewer 2: I think we all agree that the most important point to get across here is the readability/performance trade-off.  I have personally given up trying to second guess the compiler-runtime-jitter combo.  Whilst I am not questioning  @PiRXlv analysis I would rather leave the emphasis on the the trade-off.

I’m not against discussing performance and I think there are other opportunities to address this.  I usually discuss the relative performance of 2 LINQ approaches and a non-LINQ approach on Hamming and talk about my own performance tests specifying both the machine on which they are run and their unscientific nature.  I hope this conveys to the student that it’s probably best to take an evidenced-based and holistic approach to performance assessment.

The other exercise where I frequently find it necessary to draw the student's attention to performance is on Sum Of Multiples where unnecessarily slow code means that some not-very-edge cases take a disproportionate time to run but even here I leave it to the student to discover the problem and its causes.

Altogether a tricky and contentious area.

Reviewer: I agree. @PiRXlv would you be okay with removing this comment? I think it will confuse more than help to be honest.

Author: I'm sure fine with that. Actually it's slightly re-worded original text from what I remember and it was the exact reason why I opened this PR

Reviewer: No worries. Take your time.

Reviewer: @PiRXlv Small bump :)

>   * Hence, using `IEnumerable.Any()` or `IEnumerable.Where(...)` to implement the exception is wrong. It should be thrown within the single iteration.

Reviewer: ```suggestion
```

>   * Similarly, though there are numerous ways to use LINQ to solve this problem it should be avoided in performance scenarios. All of them involve iterating the input string multiple times.

Reviewer: ```suggestion
  * Similarly, though there are numerous ways to use LINQ to solve this problem, it should be avoided in performance scenarios. All of them involve iterating the input string multiple times.
```

>   * Similarly, though there are numerous ways to use LINQ to solve this problem, it should be avoided in performance scenarios. All of them involve iterating the input string multiple times.

Reviewer: ```suggestion
  * If you care about raw performance, LINQ is usually not the best option, as most of its methods involve iterating the input string multiple times.
```

> * If you look for readability oriented solution, one using LINQ probably will be better fit.

Reviewer: ```suggestion
* If one wants to optimize for readability, using LINQ is probably the best approach.
```

### Common Suggestions

As the README doesn't explicitly mention that a student should optimize for performance, don't discard solutions just on this ground. Rather, use it as a chance to start a discussion about readability/performance trade-offs and when to prefer which.

> * If you look for a performance-oriented solution, iterating over the input string more than once wastes cycles, memory, etc. Often, it isn't obvious that this is happening because methods like `ToList`, `GroupBy` and `Count` will iterate over all the string's characters without the student having to write a loop.

Reviewer: ```suggestion
* If you look for a performance-oriented solution, iterating over the input string more than once wastes cycles, memory, etc. Often, it isn't obvious that this is happening because LINQ's `ToList`, `GroupBy` and `Count` methods will iterate over _all_ the `string`'s characters.
```
